### PR TITLE
fix(#3986): a disposing parent still carries the requests its hosted hubs already accepted

### LINKED ISSUE #3986: Portal hub releases sync stream before in-flight user actions are delivered, silently discarding an accepted click
**What is failing**

A user's click on area `Catalog/Categories/Cat-Insurance` was accepted by the portal but never ran: by the time the `ClickedEvent` reached the `Store` owner hub, its per-stream `sync/{id}` sub-hub was already gone (disposed circuit / released read stream), so `DataExtensions.RefuseStreamMessage` refused it. The action did not run and will not run — the user sees nothing unless a live sender happens to exist for the `DeliveryFailure`.

**Probable cause**

High confidence, and already documented: the per-circuit portal hub released its stream while an inbound user action it had already accepted was still in flight to the owner hub — the exact ordering defect behind #3566, which the refusal mechanism deliberately does not fix. The stream ended ~5s before this line was written (the sync-hub registration grace), i.e. at circuit teardown, not at delivery time. Whether the click is lost turns on server load at the moment of the click, which explains why it is rare in testing but reachable in production. See [Refusing a Lost User Action](/Doc/Architecture/RefusingALostUserAction), which records the same class of event from CI run 34042620439.

**Impact**

Single occurrence on a single pod in this window: one silently discarded user action (the user's click produced no effect and no visible error, since their circuit had gone). Low volume, high severity-per-event — this is accepted work being thrown away.

**Where to look**

- `MeshWeaver.Data.DataExtensions.RefuseStreamMessage` — where the refusal is logged (working as designed).
- The outstanding fix lives in the Blazor portal (`MeshWeaver.Plugins`): the per-circuit portal hub should drain accepted user actions before releasing subscriptions during disposal. That cross-repo change is explicitly called out as not-yet-done in the architecture doc above.
- Related background: [Stream Liveness and the Hub Reference](/Doc/Architecture/StreamLivenessAndTheHubReference), [Error Propagation & Wedges](/Doc/Architecture/ErrorPropagationAndWedges).

Likely a duplicate in spirit of #3566's unfixed ordering half; file under the same umbrella so production occurrences of lost clicks can be tracked against the ordering fix.

---

### Evidence

| | |
|---|---|
| Fingerprint | `c3ea7263f217a7f3` |
| Category | `MeshWeaver.Data.DataExtensions` |
| Severity | Error |
| Namespace | `memex-cloud` |
| Pods | `memex-portal-deployment-5867566569-k9nmf` |
| Occurrences | 1 |
| First seen | 2026-09-10 22:08:01Z |
| Last seen | 2026-09-10 22:08:01Z |

<details><summary>Recent log lines</summary>

```
2026-09-10 22:08:01Z memex-portal-deployment-5867566569-k9nmf fail: MeshWeaver.Data.DataExtensions[0]
      REFUSING ClickedEvent on area Catalog/Categories/Cat-Insurance for stream tj8CtLoIhU225Dm_kKiYYg on hub Store: the target stream is gone (disposed circuit, released read stream, or never-created sync hub), so the action the user asked for did NOT run and never will. Sender: sync/tj8CtLoIhU225Dm_kKiYYg~portal/ITJ9qXnfLDbhctiZ_BGQb9cHLcGCJq5skaBJk3wMmYU. This line is written 5s AFTER the message arrived (the sync-hub registration grace) — the stream ended then, not now.
```

</details>

<sub>Opened automatically from `Admin/_LogIncident/c3ea7263f217a7f3`. Recurrences are folded into this issue rather than opening new ones.</sub>

### PR DESCRIPTION
**Pull request by rbuergi**
Fixes #3986

## Decided from the code, not from the live count

The question was whether an action the client **accepted** can still be lost before it is delivered on `origin/main`. Two teardown routes were driven against real hubs with the click provably still **queued** on the client-side `sync/{id}` hub when the teardown starts. The queue is parked deterministically (a `SpinWait.SpinUntil` on a volatile int, released in a `finally`), the click goes through the real `SubmitUserAction`, and the owner-side action records whether its own `sync/{id}` had already been told to go when it ran. That makes both ways of losing the click red without a timed "nothing happened" window.

| route | on `main` |
|---|---|
| the stream is released directly (navigation, eviction) | ✅ ran. #4001's release-on-the-stream's-hub holds |
| the hub **hosting** the stream is disposed (the per-circuit portal hub on circuit close) | ❌ **refused**: `Hub sync/… cannot route ClickedEvent to host/1 — its parent hub client/… is shutting down (RunLevel=DisposeHostedHubs)` |

**Root cause (measured):** a parent in `DisposeHostedHubs` refused all transit, both in the child's route-up (`HierarchicalRouting`) and at its own intake gate (tier 2 of `MessageService.RefusesIntake`). The stated reason was "the children are going down with it". But that phase is exactly when the parent *asks* its children to go down, and each child's `ShutdownRequest` queues FIFO behind the work it already accepted. So the portal hub closed its door on a click its own hosted hub had taken.

## Fix

`MessageHub.CarriesAcceptedWorkOfAHostedHub` is one predicate, and both refusal sites now ask it. It carries a delivery only when all of these hold:
- the parent is **in** `DisposeHostedHubs`
- the delivery is a request its **originating hub holds a live response callback for** (a one-way `IRequest` posted without `Observe`, and fire-and-forget, keep their drop; replies already have their own exemption)
- it is **transit**, not new work addressed to the parent
- the hosted hub handing it up is **still below `Quiescing`**. That is a structural fence: routing runs in that hub's own turn, so the delivery was queued FIFO ahead of the `ShutdownRequest` its `Dispose()` posted
- the **parent's own parent still routes**, so a whole-tree teardown is unchanged

The parent cannot reach `ShutDown` before its children complete, and a child cannot complete before handing its backlog up. So nothing new waits, and there is no timer, retry or widened bound.

## Evidence

`UserActionQueuedBehindABusySyncHubTest` (3 tests):
- the two route tests also assert that the sender's drain ended on the owner's receipt (`QuiescingTimedOut == false`), and the stream-route test asserts that the release still reaches the owner afterwards
- `AClickTakenOnAfterItsSyncHubWasAskedToGoDownIsRefusedNotCarried` pins the other side of the fence (added after the Copilot review): a click queued *behind* the sync hub's own `ShutdownRequest` is refused, not carried
- every park asserts it released on its intended condition rather than on its budget

| mutant | result |
|---|---|
| `main` (no fix) | client-hub test red (`refused: … cannot route ClickedEvent … RunLevel=DisposeHostedHubs`); stream test green |
| release registered on the STREAM again (#4001 reverted) | stream test red (`ran on an owner-side handler already told to go`), `AnAcceptedActionHoldsTheReleaseUntilTheOwnerAnswers` red; client-hub test green |
| only the parent-intake half removed | client-hub test red (`Hub client/… is shutting down … cannot process ClickedEvent`) |
| only the child route-up half removed | client-hub test red (`cannot route ClickedEvent …`) |
| "still below `Quiescing`" clause dropped | fence test red (`Expected "ran" to start with "refused"`); both route tests green |

Local, Debug, final head: `MeshWeaver.Messaging.Hub.Test` 461/461, `MeshWeaver.Layout.Test` 507/507. Teardown subsets, also on the final head: Teardown subsets: Graph.Test 17/17 (`NackReachesTheWaiterDuringTeardown`, `ReplyRoutedDuringMeshTeardown…`, `RefusedReplyReachesTheWaiter`, `UndeliverableReplyShapes`, `CreateAnswersWhenTheOwnerGoesAway`, `LateNackReenqueue`, `ARecyclingOwnerIsNotAnUnknownRefusal`), Hosting.Test 10/10, Data.Test 37/37. Release `-warnaserror` (final head): `MeshWeaver.Messaging.Hub`, `MeshWeaver.Layout.Test`, `MeshWeaver.Documentation.Test` each 0 warnings / 0 errors; `MeshWeaver.Documentation.Test` 496/496.

## Measured on the same route, NOT changed here

With nothing owed, disposing the client hub does **not** release the owner-side `sync/{id}`: it had not disposed after 12 s. The `UnsubscribeRequest` is posted by the parent from the child's `ShutDown`, while the parent is in `DisposeHostedHubs`, and its own fire-and-forget post is refused at intake. So the Plugins `CircuitAccessHandler` comment ("each child's teardown posts UnsubscribeRequest to its owner node, so the owner-side mirrors close too") does not hold on that route. It loses no click, so it is recorded in `Doc/Architecture/RefusingALostUserAction` as a candidate for #3432 and not folded in here.

## Docs

- `Doc/Architecture/RefusingALostUserAction`: new section with the measurement, fix and falsifications
- `Doc/Architecture/TeardownLayers`: the tier-2 transit carve-out
- What's New (`Category: Fix`)

Pairs-with: none — no public type or member removed; the new members are internal/private.
Mirror-sync: none — no i18n key added or changed.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/MeshWeaver.Messaging.Hub/MessageHub.cs [hunk 1]
```diff
@@ -1733,6 +1733,82 @@ private System.Reactive.Subjects.AsyncSubject<IMessageDelivery> GetOrAddResponse
...
+    internal bool CarriesAcceptedWorkOfAHostedHub(IMessageDelivery delivery)
+    {
+        if (RunLevel != MessageHubRunLevel.DisposeHostedHubs
+            || delivery.Properties.ContainsKey(PostOptions.RequestId)
+            || !MessageService.IsAwaitedBySender(delivery)
+            || delivery.Target is not { } target
+            || (target with { Host = null }).Equals(Address)
+            || (messageService as MessageService)?.ParentHub is not { RunLevel: < MessageHubRunLevel.DisposeHostedHubs })
+            return false;
+        return HostedHubThatSent(delivery.Sender) is { RunLevel: < MessageHubRunLevel.Quiescing };
```
[the remark this answers is not in the document]
Comment (rbuergi) [reply src/MeshWeaver.Messaging.Hub/MessageHub.cs]: I've addressed this in the follow-up commit. I tightened the fence and added the missing regression test, but the fence is the child's own `ShutdownRequest`, not the parent's phase boundary.

**The fence is structural.** Routing up runs inside the child's own turn. The child leaves `Started` only by handling the `ShutdownRequest` that its `Dispose()` posts, FIFO. So a delivery routed while the child is still below `Quiescing` was queued ahead of that request, which means it was accepted before the child's teardown. The window you describe is a post landing in a still-`Started` child after the parent flips to `DisposeHostedHubs` but before that child's `Dispose()`. That window sits inside the parent's single synchronous `hostedHubs.Dispose()` call. A delivery caught in it is still ahead of the child's `ShutdownRequest`, so by the teardown contract it is accepted work. Each hub's own tier-1 gate owns the no-new-work guarantee, and that gate is unchanged.

**The regression test for the other side is added.** `AClickTakenOnAfterItsSyncHubWasAskedToGoDownIsRefusedNotCarried` waits until the client hub's `DisposeHostedHubs` has disposed the sync hub and that hub's `ShutdownRequest` is provably in its queue. Only then does it submit the click, from the parked turn, and asserts the click is **refused**. Dropping the "still below `Quiescing`" clause turns it red: `Expected "ran" to start with "refused"`. The two route tests stay green under that change.

**From the suppressed findings, also changed:**
- The predicate now requires a live response callback on the originating hub (resolved down the sender's host chain) instead of the type-level `IsAwaitedBySender`. A one-way `IRequest` posted without `Observe` keeps its drop. `IsAwaitedBySender` is private again.
- Every park now records whether it released on its intended condition or on its budget, and each test asserts the former.

**Not changed:** the `RouteAddressToHub` / `RouteAddressToHostedHub` gates. They route into hubs this parent hosts, and those hubs are all being disposed in this phase. Whether a request should still be handed to such a sibling is a different question from carrying a child's accepted work out. #3986's route does not pass through those gates: the portal hub's only configured route is for kernel addresses, and the click leaves through `HierarchicalRouting`. I left this out of scope on purpose.
